### PR TITLE
Math functions for Swift

### PR DESCRIPTION
**Pull request by stephentyrone**
This PR implements SE-0246.

We still need to make a decision about how to handle obsoleting functions provided by the platform module. Because most programs pull in Foundation, they transitively pull in the existing names, which means that most functions calls in concrete contexts will get the tgmath functions, not the stdlib implementations. That's OK, but there are a few functions for which the behavior differs for edge cases (most notably `pow`), and we'd also like to remove these to cut down the overload sets.

<rdar://problem/17625344>

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by tkremenek** [comment]
@swift-ci clean test

**Comment by tkremenek** [comment]
@stephentyrone @airspeedswift what is preventing this from moving forward?

**Comment by stephentyrone** [comment]
@tkremenek Just shaking out fallout from dropping the Math module, specifically what that means for the existing Platform functions. I expect to push this forward today.

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci clean test

**Comment by stephentyrone** [comment]
@swift-ci please smoke test

**Comment by stephentyrone** [comment]
@swif-ci please smoke test.

**Comment by stephentyrone** [comment]
@swift-ci please smoke test.

**Comment by stephentyrone** [comment]
@swift-ci please test.

**Comment by stephentyrone** [comment]
@swift-ci please test.

**Comment by compnerd** [comment]
This broke the windows build! https://dev.azure.com/compnerd/windows-swift/_build/results?buildId=976

**Comment by stephentyrone** [comment]
@compnerd windows doesn't have `hypot` either?!

**Comment by compnerd** [comment]
@stephentyrone - clearly MSVCRT hates math :-(

```
Swift.obj : error LNK2019: unresolved external symbol hypotf referenced in function $sSf5hypotyS2f_SftFZ
```

I think it may be inlined or something, since they claim it is in the library:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/hypot-hypotf-hypotl-hypot-hypotf-hypotl?view=vs-2019

**Comment by stephentyrone** [comment]
@compnerd Can you replace _swift_stdlib_hypotf with a bogus stub on a branch and see if anything else breaks?

**Comment by compnerd** [comment]
@stephentyrone - same error :/

**Comment by compnerd** [comment]
Yeah, guess we need to switch it over to `_hypotf`:

```
    _Check_return_ __inline float __CRTDECL hypotf(_In_ float _X, _In_ float _Y)
    {
        return _hypotf(_X, _Y);
    }
```

I'll put up a patch momentarily.

File: stdlib/public/core/MathFunctions.swift.gyb [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+/// Conformance to this protocol means that all of these building blocks are
+/// available as static functions on the type.
+///
+/// ```swift
+/// let x: Float = 1
+/// let y = Float.sin(x) // 0.84147096
+/// ```
+///
+/// All of these are made available as free functions by importing the Math
+/// module:
```
Comment (xwu) [inline stdlib/public/core/MathFunctions.swift.gyb]: This is outdated, I thought?

Comment (stephentyrone) [reply stdlib/public/core/MathFunctions.swift.gyb]: Yeah, I still need to update some comments.

**Pull request merged**